docs/okeps/okep-5233-pre-assigned-pod-net-id.md

```diff
@@ -0,0 +1,455 @@
+
+
+The following scenarios should be covered in testing:
+
+* VM workloads import into OVN-Kubernetes with no changes to the instances network configuration.
+* Imported VM workloads can live-migrate to another node without any additional traffic disruption.
+* 'v1.multus-cni.io/default-network' cannot be changed after the pod was created.
+* It should be possible to configure the pods MAC or the IP address without configuring the other.
+* When `excludeSubnets` is configured automatic IP allocation should not use addresses specified in it.
+* It should be possible to configure the pods IP address using the 'v1.multus-cni.io/default-network'
+even if the address is a part of the `excludeSubnets`.
+* Requesting an IP address that is not a part of the networks subnet should fail.
+* Detect MAC and IP address conflicts between the requested addresses for a newly created pods and the addresses that
+are already allocated in the network.
+* After configuring custom default gateway and management addresses on a Layer2 UDN the previous default
+IPs can be consumed by workloads(e.g. for 10.0.0.0/16 network create pods with 10.0.0.1 and 10.0.0.2 addresses).
+* Modifying the  default gateway and management addresses on a Layer2 UDN should not be possible after the network
+was created.
+
+The scenarios mentioned above have to cover both IPv4 and IPv6 IP families.
+
+
+### Documentation Details
+
+## Risks, Known Limitations and Mitigations
+
+* Modifying the 'v1.multus-cni.io/default-network' value after the pod was created could have unpredictable
+consequences.
+To mitigate this introduce a Validating Admission Policy described in [Implementation Details](#implementation-details).
+
+* By allowing users to specify the IP and MAC addresses for the pods there is a risk of conflicts.
+To mitigate this OVN-Kubernetes will check that the requested addresses are not currently used in the UDN.
+There is still a risk that the user picks an address that's consumed by something outside of the UDN but that's beyond
+what OVN-Kubernetes controls and can check.
+
+* The dynamic, per-node subnet allocation in Layer3 UDNs, where each node has a unique default gateway and
+management IP, makes user-specified UDN gateway/management IPs and static pod IP/MAC assignments very complex. This
+enhancement will not support Layer3 UDNs.
+
+* BGP support today is limited to cluster UDNs, to ensure a non-NATed traffic for pods with pre-configured addresses
+the user has to use a cluster UDN to configure the network. This is a limitation unrelated to this enhancement
+and it is possible it will be solved in the future.
+
+* By consuming the 'v1.multus-cni.io/default-network' annotation for altering the primary UDNs pod configuration the
+user won't be able to use it for configuring the cluster default network attachment. This is acceptable as there is
+currently no support for modifying the cluster default network through this annotation while using primary UDNs.
```

Comment:
https://docs.redhat.com/en/documentation/openshift_container_platform/4.14/html/networking/hardware-networks#adding-network-attachment-definition-to-pods_configuring-hardware-offloading
I guess we break hardware offload semantics for a pod cause this annotation wasn't only about "cluster default network" right? it was about allowing the pod to "have whatever it wants as its primary network" but I guess you are right that with PUDNs we never really supported this - though adding support for offload on UDN network itself was something someone might ask for but we now have yet another feature that leverages this. good to call it it out that in future i dunno if support for both can be considered.

Reply:
I don't think we are breaking it since it is not there today but that was the purpose of this point yeah, to highlight we are taking over the annotation for configuring the primary network.

Reply:
let me re-phrase: we are not locking ourselves from being able to support a future case of supporting sriov offload with udn pods right? that can still be done and that and this feature can co-exist correct?